channeldb/kvdb/etcd/readwrite_bucket.go

```diff
@@ -0,0 +1,246 @@
+package etcd
+
+import (
+	"strconv"
+	"strings"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+type readWriteBucket struct {
+	path string
+	tx   *readWriteTx
+}
+
+func newReadWriteBucket(tx *readWriteTx, path string) *readWriteBucket {
+	return &readWriteBucket{
+		path: path,
+		tx:   tx,
+	}
+}
+
+// NestedReadBucket retrieves a nested bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadBucket(key []byte) kvdb.ReadBucket {
+	return b.NestedReadWriteBucket(key)
+}
+
+// ForEach invokes the passed function with every key/value pair in
+// the bucket.  This includes nested buckets, in which case the value
+// is nil, but it does not include the key/value pairs within those
+// nested buckets.
+func (b *readWriteBucket) ForEach(cb func(k, v []byte) error) error {
+	prefix := makeValuePrefix(b.path)
+	prefixLen := len(prefix)
+
+	kv := b.tx.stm.First(prefix)
+	for kv != nil {
+		if err := cb([]byte(kv.key[prefixLen:]), []byte(kv.val)); err != nil {
+			return err
+		}
+		kv = b.tx.stm.Next(prefix, kv.key)
+	}
+
+	prefix = makeBucketPrefix(b.path)
+	prefixLen = len(prefix)
+
+	kv = b.tx.stm.First(prefix)
+	for kv != nil {
+		if err := cb([]byte(kv.key[prefixLen:]), nil); err != nil {
+			return err
+		}
+		kv = b.tx.stm.Next(prefix, kv.key)
+	}
+
+	return nil
+}
+
+// Get returns the value for the given key.  Returns nil if the key does
+// not exist in this bucket (or nested buckets).
+func (b *readWriteBucket) Get(key []byte) []byte {
+	if len(key) == 0 || !isValidKey(key) {
+		return nil
+	}
+
+	valueKey := makeValuePrefix(b.path) + string(key)
+	val := b.tx.stm.Get(valueKey)
+	if val == "" {
+		return nil
+	}
+
+	return []byte(val)
```

Comment:
similarly we may have to decode hex/base64 back into binary if we needed to convert it when storing.

Reply:
Not sure I understand what you meant here. If you referred to binary keys, that's done.